### PR TITLE
Enable code generation features in tool (via opt-in)

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Allows usage of build_runner builders as an integrated part of flutter's test, hot reload, and build workflows. This feature is a developer preview, I expect that there will be follow up fixes and bugs required, and their may need to be changes to configuration or versions accepted. 

Example: to use json_serializable in a project:

pubspec.yaml
```
name: example_project

dependencies:
  json_annotation:  ^2.0.0

builders:
  json_serializable: ^2.0.3
```

After `flutter run` or `flutter generate`, the generated code will appear in the build cache. The editor may need to be restarted if this is the first run of code generation.

Currently we only support building to cache for the sake of our hot reload integration. We'll need to revisit this and provide guidance for package publishing.

## Related Issues

https://github.com/flutter/flutter/issues/27141

Requires https://github.com/flutter/flutter/pull/28152

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jonahwilliams** [comment]
cc @yjbanov

File: packages/flutter_tools/lib/src/compile.dart [hunk 1]
```diff
@@ -16,11 +16,24 @@ import 'base/io.dart';
...
+import 'project.dart';
 
-KernelCompiler get kernelCompiler => context[KernelCompiler];
+KernelCompilerFactory get kernelCompilerFactory => context[KernelCompilerFactory];
+
+class KernelCompilerFactory {
+  const KernelCompilerFactory();
+
+  Future<KernelCompiler> create(FlutterProject flutterProject) async {
+    if (flutterProject == null || !flutterProject.hasBuilders) {
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/compile.dart]: Is it possible for a project to go from not having builders to having some while a `flutter` command is running? If so, is this case handled? IOW, would the factory be rerun to get the code-generating compiler?

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 2]
```diff
@@ -997,11 +999,27 @@ class ProjectFileInvalidator {
         printError('Warning: updated dependencies detected. The Flutter application will require a restart to safely use new packages.');
       }
       _packagesUpdateTime = newPackagesUpdateTime;
+      final bool hadBuilders =_hasBuilders;
+      _hasBuilders = _flutterProject?.hasBuilders ?? false;
+      if (hadBuilders != _hasBuilders) {
+        printError('Warning: Added builders detected. The Flutter runner will require a restart to use them.');
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/run_hot.dart]: Ah, looks like the case I'm describing above is handled here.

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 3]
```diff
@@ -997,11 +999,27 @@ class ProjectFileInvalidator {
...
-      _scanDirectory(packageUri, invalidatedFiles);
+      if (_flutterProject != null && _flutterProject.manifest.appName == packageName && _hasBuilders) {
+        final Uri generatedUri = _flutterProject
+          .generated
+          .childDirectory('lib')
+          .uri;
+        _scanDirectory(packageUri, invalidatedFiles);
+        _scanDirectory(generatedUri, invalidatedFiles);
+      } else {
+        _scanDirectory(packageUri, invalidatedFiles);
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/run_hot.dart]: Looks like you could avoid the `else` clause by moving `_scanDirectory(packageUri, invalidatedFiles);` above the `if` block. It has to run either way.

**Review by yjbanov — APPROVED** [review]
lgtm

File: packages/flutter_tools/lib/src/compile.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/compile.dart]: Yes,  we won't be able to switch them currently. This does fall under the umbrella for "hot reload literally everything" that I am working on.

The solution is to: 1) always run with build_runner enabled and 2) use manifest invalidation as triggers to update the build script

However, that is a much bigger change. I don't think it should block letting people use the feature

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/run_hot.dart]: I've removed this due to the compiler refactor

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/run_hot.dart]: This is no longer necessary

File: packages/flutter_tools/lib/src/compile.dart (hunk 1 above)
Comment (yjbanov) [reply packages/flutter_tools/lib/src/compile.dart]: Makes sense.

File: packages/flutter_tools/test/tester/flutter_tester_test.dart [hunk 4]
```diff
@@ -194,3 +195,14 @@ Hello!
...
+class FakeKernelCompilerFactory implements KernelCompilerFactory {
+  FakeKernelCompilerFactory(this.kernelCompiler);
+
+  final KernelCompiler kernelCompiler;
+
+  @override
+  Future<KernelCompiler> create(FlutterProject flutterProject) async {
+    return kernelCompiler;
+  }
+
```
Comment (yjbanov) [inline packages/flutter_tools/test/tester/flutter_tester_test.dart]: nit: unnecessary empty line

**Review by yjbanov — APPROVED** [review]

**Pull request merged**